### PR TITLE
Support PSR-3 logging

### PR DESCRIPTION
**Pull request by jacobkiers**
Hi Marcelo,

Since most of the PHP world (including us) moved to PSR-3 during the last few year, I thought it worthwhile to add PSR-3 support to PAMI.

I've maintained backwards compatibility by using an adapter.

What do you think?

### CODE REVIEW

File: src/mg/PAMI/Logger/Log4PhpAdapter.php [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    {
+        $this->log(PsrLevel::ALERT, $message, $context);
+    }
+
+    public function critical($message, array $context = array())
+    {
+        $this->log(PsrLevel::CRITICAL, $message, $context);
+    }
+
+    public function error($message, array $context = array())
```
Comment (marcelog) [inline src/mg/PAMI/Logger/Log4PhpAdapter.php]: I think the log levels are wrong for debug, info, notice, warning, and error

File: src/mg/PAMI/Client/Impl/ClientImpl.php [hunk 2]
```diff
@@ -185,9 +189,7 @@ public function open()
 	    @stream_set_blocking($this->_socket, 0);
 	    $this->_currentProcessingMessage = '';
 	    //register_tick_function(array($this, 'process'));
-	    if ($this->_logger->isDebugEnabled()) {
```
Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: The check is done at this level so we can avoid (if possible) the string concatenation and all the work that implies

File: src/mg/PAMI/Logger/Log4PhpAdapter.php (hunk 1 above)
Comment (jacobkiers) [inline src/mg/PAMI/Logger/Log4PhpAdapter.php]: Oops! Copy-paste error. I'll fix it in the next commit. Do  you prefer a rebase (keeps the history clean), or a new commit (which makes it easier to spot the changes)?

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 2 above)
Comment (jacobkiers) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: I see. Sadly, PSR-3 does not support that.

We have personally been running this in production for the last 2 years or so. The actual overhead is in writing to the log file though, not the string concatenation.

Is this a blocking issue for you?

Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: hmm.. well there *is* a lot of debug output, and we can save some cpu cycles right there, since we can't decouple the logging from the main loop.

File: src/mg/PAMI/Logger/Log4PhpAdapter.php (hunk 1 above)
Comment (marcelog) [inline src/mg/PAMI/Logger/Log4PhpAdapter.php]: Rebase sounds good :)

File: src/mg/PAMI/Client/Impl/ClientImpl.php [hunk 3]
```diff
@@ -443,11 +439,14 @@ public function close()
 	 */
 	public function __construct(array $options)
 	{
-        if (isset($options['log4php.properties'])) {
-            \Logger::configure($options['log4php.properties']);
-        }
-        $this->_logger = \Logger::getLogger('Pami.ClientImpl');
-	    $this->_host = $options['host'];
+		$this->setLogger(
+			new Log4phpAdapter(
```
Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: I'm not sure about hardcoding this here. Either we should be injecting the logger by constructor or just leave things as they are, since the logger remains completely internal to pami. Or maybe I'm missing something?

Comment (jacobkiers) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: I added this to keep full backwards compatibility. If we require people to inject the logger, we are breaking that, which makes for unhappy users.

Personally, I usually try to adhere quite strictly to [semver](http://semver.org/), because it guarantees BC for minor releases. In that case, another options is to *do* require this in the constructor, and then to bump PAMI to version 2.0.0. But that seems a bit overkill to me.

As an aside: I don't see much difference between hardcoding `\Logger` and hardcoding `\Log4phpAdapter` :smile:

**Comment by jacobkiers** [comment]
By the way: I will probably do the same for PAGI, if you're interested.

File: src/mg/PAMI/Logger/Log4PhpAdapter.php (hunk 1 above)
Comment (jacobkiers) [inline src/mg/PAMI/Logger/Log4PhpAdapter.php]: Done.

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 3 above)
Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: We can make it an "optional option", and not break older implementations, allowing users to use other loggers, keeping the old behavior available when the new option is not available (just thinking out loud).

The hardcoding was done because there was not a standard interface for loggers, and log4php offered some cool features that loggers didn't offer at the time, so hardcoding the logger made more sense than writing an arbitrary logger interface not used by others in the community.

Since in this case we are dealing with a PHP standard, it makes more sense to use that as an interface and have the user provide the implementation, without making pami hardcoding a specific implementator.

From my point of view, writing an adapter and hardcoding it here seems to accomplish the same as leaving things as they are, since users can't change implementations.

I know, the user can later on change the implementation via setter injection, but it will force the user to momentarily use log4php (when instantiating the client) until the logger is set.

Just pointing out these details.

**Comment by marcelog** [comment]
I really like the idea :) Since I'm a bit out of the php world I'd like to read more about PSR3 first, and then play a bit with your changes. If this gets merged, I would really like to have in PAGI too, of course :)

Thank you very much for the effort and the contribution!

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 3 above)
Comment (jacobkiers) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: I agree, and in our own implementation we set the logger using the `$config['logger']` parameter.

How about this: I do the same here, but if the `$config['logger']` property is not set, we keep the behaviour as proposed in this PR?

I think this satisfies both cases: existing users don't have to change anything, and users who want to use PSR-3 can just inject a logger using `$config['logger']`.

File: src/mg/PAMI/Client/Impl/ClientImpl.php [hunk 4]
```diff
@@ -443,11 +439,14 @@ public function close()
 	 */
 	public function __construct(array $options)
 	{
-        if (isset($options['log4php.properties'])) {
-            \Logger::configure($options['log4php.properties']);
-        }
-        $this->_logger = \Logger::getLogger('Pami.ClientImpl');
-	    $this->_host = $options['host'];
+		$this->setLogger(
```
Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: Hmm.. we shouldn't be calling an instance method from the constructor

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 2 above)
Comment (jacobkiers) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: I have given this some thought. I *can* write a series of adapters specific for PAMI in order to solve this. However, the design would be kind of ugly and quite a bit complicated.

TL;DR: The costs are high and the winnings are (relatively) low. I would strongly advice *against* re-introducing this behaviour. The reasons are:

1. It directly contradicts the goals of PSR-3, which is to provide a standardised logging solution.
1. In the default case, we don't win anything over the current situation.
1. The complexity of the implementation is high.
1. The complexity of documenting this behaviour is high.
1. Finally, these CPU cycles can be won back (and then some) by upgrading to the latest PHP versions. These usually have so many performance improvements that the cost of having these calls are dwarfed by it. Since we are targeting people who run Asterisk, we have excluded almost by definition the users on shared hosting. Therefore, the remaining users will usually be able to do such an upgrade.

### Proposed implementation

Just so you know that I don't randomly shout that it's a bad idea (:wink:), here is the implementation that I think is required to re-introduce the behaviour as it was.

We will need two extra marker interfaces (don't bother with the names just yet...) and another adapter:

 * `ILogLevelEnabledChecker`
 * `ILogLevelDisabler`
 * `PAMILoggerAdapter` (implements `Psr\Log\LoggerInterface` and `Psr\Log\LoggerAwareInterface`)

We need both, because Log4Php already has the functionality to enable and disable log levels (this not needing `ILogLevelDisabler`. Also, the `Log4PhpAdapter` needs to implement `ILogLevelEnabledChecker`.

Specific implementations of PSR-3, such as Monolog, have this functionality also, but we cannot count on that as we are programming to an interface.

We will need the `PAMILogger` adapter to convert a generic PSR-3 Logger into one that has these two interfaces. This adapter must have the default to be that it enables debug logging, otherwise it the results will be very surprising to users. A user that *does* want to disable specific log levers, must wrap it in `PAMILogger` by himself, and disable these log levels specifically.

Furthermore, we will need to allow both plain PSR-3 Loggers (which will then get wrapped in `PAMILogger` and instances of `ILogLevelEnabledChecker` (which will not be wrapped).

Finally, we have to re-introduce the checks of whether these log levels are enabled or disabled.

All in all, even though I *do* understand the reasons of wanting this functionality, I don't think that it warrants the cost, both in complexity of implementation and in complexity of documentation. In the default case (when a user just injects a PSR-3 logger), it will still need to do the concatenation anyway, because we cannot know the intention of the user upfront.

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 4 above)
Comment (jacobkiers) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: Excuse me? Could you please explain why? At least in the PHP world, calling setters from the constructor is perfectly acceptable.

One should not do actual *work* from the constructor, but validating parameters, instantiating objects (smelly, but OK) and setting properties is not a problem.

**Comment by jacobkiers** [comment]
FYI, you can read more about PSR-3 in the specification: http://www.php-fig.org/psr/psr-3/.

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 4 above)
Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: Setting a property is ok, but calling an instance method in a constructor is smelly, and probably a bad practice overall

Comment (jacobkiers) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: If you really care, I will set it directly.

I'd much rather like you to review and respond to [this comment](#discussion_r28460103). If you would be so kind :)

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 2 above)
Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: I think you went a bit too far with that :) Differences aside, the important matter is that the concatenations might be avoided by using the logger context, and passing the needed key/values (although we have to trust that the logger implementation will be reasonable enough as to avoid any further processing when the log level is not appropiate). It's a bummer that PSR3 wont even return the current loglevel.

I'm not sure about

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 3 above)
Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: That would be fine

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 2 above)
Comment (jacobkiers) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: I don't understand what you are trying to say here. And it looks like some part of your comment got lost. 

The logger context parameter in PSR 3 has nothing to do with the log level. It is used to provide some meta data to the logger. Or I misunderstand your use of the word context here. 

Could you paste some sample code in the comment to help me understand?

Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: Concatenation is just a way of string interpolation. So instead of doing something like `"Note: ". $msg . " something else"` the logger can be called with a variable `message = value` and the log string would be `Note: {message} something else`

Comment (jacobkiers) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: Well, that's not what a standard PSR 3 logger does either :). It's a very basic interface. But I guess we can change the message building to `"Note {$variable} something"` syntax. Is that what you mean?

**Comment by jacobkiers** [comment]
It's been a long day for me. I'll work on it again tomorrow or Friday. Thanks for the quick responses so far.

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 2 above)
Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md#12-message

Using the context effectively avoids the concatenation (provided that the logger implementation is smart enough to replace the placeholders only after checking the log level)

Comment (jacobkiers) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: Nice catch! I didn't even remember that :-). Let's try it!

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 3 above)
Comment (marcelog) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: Let's call it "psr3_logger" so it is perfectly clear the kind of object expected

**Comment by jacobkiers** [comment]
Hi @marcelog, I've updated both with your comments.

As for the logger injection, I've chosen (for now) to keep using the setter. This enforces that the given logger is indeed an instance of `LoggerInterface` without me having to check that manually in the constructor.

However, as I said before, if you really do care, I will update it so that it is set directly on the property.

File: src/mg/PAMI/Client/Impl/ClientImpl.php (hunk 2 above)
Comment (dkgroot) [inline src/mg/PAMI/Client/Impl/ClientImpl.php]: > 5. Finally, these CPU cycles can be won back (and then some) by upgrading to the latest PHP versions. These usually have so many performance improvements that the cost of having these calls are dwarfed by it. Since we are targeting people who run Asterisk, we have excluded almost by definition the users on shared hosting. Therefore, the remaining users will usually be able to do such an upgrade.

FYI: You might be shoked, but some FreePBX users seem to stuck in a timewarp of sorts, for example people using Elastix 2.5 (Centos 5.5 / Asterisk-1.8 / php 5.1 / non patched bash / nasty openssl). Upgrading php could be very problematic for them. 

I think they are nuts, running such old version of "everything", but they are out there. This is just to counter your assumption that not running in a hosted environment means up to date / updatable. This does not mean i would like to hold you back from using/implementing PSR3 and making people move forward.

File: composer.json [hunk 5]
```diff
@@ -20,6 +20,7 @@
   },
   "require": {
     "php": ">=5.3.3",
-    "apache/log4php": "2.3.0"
+    "apache/log4php": "2.3.0",
```
Comment (johanwilfer) [inline composer.json]: Is log4php really a "require"? Maybe a "suggest"?
https://getcomposer.org/doc/04-schema.md#suggest

**Comment by marcelog** [comment]
@jacobkiers I want to merge this one, but at the same time would like to remove all PEAR references (for documentation, examples, and tests since PSR3 is only available via composer), and also fix some needed stuff (examples need to use the autoload.php file for ps3, etc). Would you be so kind to open this pull request **as it is** against the `1.80.0` branch instead of master? We can merge to master once I've done those changes.

Thanks for staying with me so far!

File: composer.json (hunk 5 above)
Comment (jacobkiers) [inline composer.json]: @johanwilfer Yes, it was to maintain backwards compatibility. However, that may be fixed in the 1.80.0 branch as Marcelo mentioned when closing this PR.

Thanks for the review though!

**Comment by jacobkiers** [comment]
@marcelog Sure!

File: composer.json (hunk 5 above)
Comment (marcelog) [inline composer.json]: @jacobkiers @johanwilfer how would that work with an "out of the box" installation? If one does not specify a PSR3 logger with the new option, [there's a default implementation](https://github.com/marcelog/PAMI/pull/83/files#diff-619e399f4b58646d77fb4f82900c73dbR445) so backwards compatibility is mantained and log4php is used. If it defined as a "suggested" package, we would have to declare all PSR3 implementors, just to be fair with them. I think we should declare it as required until (let's say) v2.0.0. 

Thoughts?

Comment (jacobkiers) [inline composer.json]: @marcelog You're right, and I agree.

I kind of forgot about the out of the box installation. I only use PAMI as part of larger projects, in which it is integrated as a library. Thanks for mentioning the standalone use case!

For v2.0.0 I would propose using [monolog](https://github.com/seldaek/monolog) out of the box. I have a few reasons for that:

* Monolog is by far the [most used logger for PHP](https://packagist.org/search/?q=psr-3&orderBys[0][sort]=downloads&orderBys[0][order]=desc) these days.
* It fully supports PSR-3.
* It provides a staggering amount of logging protocols and services out of the box, making it very easy to log stuff just the way the user wants.

But for now, I'm happy with it as-is.

Comment (marcelog) [inline composer.json]: Perhaps for v2.0.0 we can get rid of any "default" logger, and just clearly state that the user should add one by him/herself. No biggie, since we're going to do a major version change. And it's not unreasonable to ask for a logger as the only dependency (out of the dev environment, of course).

File: src/mg/PAMI/Logger/Log4PhpAdapter.php [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+namespace PAMI\Logger;
+
+use Logger;
+use LoggerLevel as Log4PhpLevel;
+use LoggerLevel;
+use Psr\Log\LoggerInterface;
+use Psr\Log\LogLevel as PsrLevel;
+
+class Log4PhpAdapter implements LoggerInterface
```
Comment (marcelog) [inline src/mg/PAMI/Logger/Log4PhpAdapter.php]: @jacobkiers I was just thinking.. would you like to have this adapter as a separate project and include it as a dependency? Or maybe even better: open a pull request to the log4php guys. I just thought about this because we would need to duplicate this code for pagi

File: composer.json (hunk 5 above)
Comment (johanwilfer) [inline composer.json]: I don't think you need to add all PSR3 implementors as "suggested" just to be fair. If you wan't you could require the "psr/log" package that holds the interface and all implementations "provide" this interface: https://packagist.org/packages/psr/log

But as you say, this could be optional for v2.0.0. If a logger is provided it is used, if not you don't nothing is logged.
Monolog could be "suggesed", as this is a very competent logger.

File: src/mg/PAMI/Logger/Log4PhpAdapter.php (hunk 6 above)
Comment (jacobkiers) [inline src/mg/PAMI/Logger/Log4PhpAdapter.php]: Given the fact that the log4php project hasn't had any commits in the past two years, I don't see much use in creating a PR for this.

Splitting this adapter out into a separate package on the other hand, would certainly be possible though.
